### ISSUE TITLE
Differences between RHSCL and openshift MongoDB images

### ISSUE THREAD

**Issue by omron93** [sclorg/rhscl2dockerfile#21]
This issue should track what are the differences between openshift mongodb image available at https://github.com/openshift/mongodb and RHSCL image generated here.

**Comment by mfojtik** [comment]
> RHSCL image doesn't have replication -> same discussion as for other databases

We are not requiring replication to be part of the image. But you have to think that people will use the this image to create replicated setups, which have to be taken into account.

> RHSCL image does not include test directory (https://github.com/openshift/mongodb/tree/master/2.4/test). rh-mongodb26 includes rh-mongodb26-mongodb-test package which includes upstream test suite

We require tests for MongoDB image, not for the mongodb26 RPM. We need to make sure that the MongoDB is started properly, it has users created, database initialized and all configuration set. Would be nice to execute the upstream test package against the container inside the 'test/run' script.

> RHSCL mongodb doesn't hadle MONGODB_NOPREALLOC, MONGODB_SMALLFILES, MONGODB_QUIET - now RHSCL mongodb accepts MONGOD_CONFIG_option and MONGOS_CONFIG_option variables and updates/adds options in mongod.conf or mongos.conf -> also the same as for other databases

Yeah, and we had weeks of conversations about this already, so instead of starting a new one here, we have to wait for PM to decide what we should do.

> RHSCL use variables "about creating users" in different way - see https://github.com/sclorg/rhscl-dockerfiles/blob/master/rhel6.rh-mongodb26/README.md#authentication

What is the difference, it seems like we contain the same set of variables.

**Comment by omron93** [comment]
> We are not requiring replication to be part of the image. But you have to think that people will use this image to create replicated setups, which have to be taken into account.

I know. It is possible by mounting scripts into (pre/post)init...

>We require tests for MongoDB image, not for the mongodb26 RPM. We need to make sure that the MongoDB is started properly, it has users created, database initialized and all configuration set. Would be nice to execute the upstream test package against the container inside the 'test/run' script.

I don't know much about running tests in openshift. Where you use them?

@hhorak What about RHSCL and tests?

>Yeah, and we had weeks of conversations about this already, so instead of starting a new one here, we have to wait for PM to decide what we should do.

I absolutely agree.

> What is the difference, it seems like we contain the same set of variables.

RHSCL MongoDB by default (without any variables) starts without auth. You can enable it with MONGODB_AUTH=true (no user creation - useful with mounted database with initialized users) or by setting MONGODB_ADMIN_PASSWORD=xxx (creates admin user and enable auth). Additionally to MONGODB_ADMIN_PASSWORD you can also specify one user you want to create...

OpenShift mongodb24 aren't able to run without auth or enable auth without creating users (MONGODB_USER + MONGODB_PASSWORD). If you want to create new user you don't have to specify database in which to create it (username by default). Also you don't have to specify (MONGODB_ADMIN_PASSWORD) and mongodb generate admin password and print it to stdout.

**Comment by hhorak** [comment]
@mfojtik wrote:
> @omron93 wrote:

>> RHSCL image doesn't have replication -> same discussion as for other databases

> We are not requiring replication to be part of the image. But you have to think that people will use this image to create replicated setups, which have to be taken into account.

The solution here is to have a directory where people will place scripts into to do all the complex stuff for setup master/slave when creating another layer  -- so users are able to create another layer **easily**.

**Comment by hhorak** [comment]
@omron93 wrote:
> @mfojtik wrote:

>> We require tests for MongoDB image, not for the mongodb26 RPM. We need to make sure that the MongoDB is started properly, it has users created, database initialized and all configuration set. Would be nice to execute the upstream test package against the container inside the 'test/run' script.

> I don't know much about running tests in openshift. Where you use them?
@hhorak What about RHSCL and tests?

The tests used in openshift are not part of the image, so we don't have to care about them right now -- we'll use https://github.com/Containers-Testing-Framework for testing images in RHSCL later. In order to keep image as small as possible, I think we don't want to install `rh-mongodb26-mongodb-test` into the image. Those tests are run during RPM build, which is fine enough.

**Comment by omron93** [comment]
>>    We are not requiring replication to be part of the image. But you have to think that people will use this image to create replicated setups, which have to be taken into account.

>I know. It is possible by mounting scripts into (pre/post)init...

Or it is also possible when you have mounted database (to persist container restart). And you can start container with some options, connect to it with mongo shell, run something, restart mogodb container with different options, connect with mongo shell and initiate replicaset. (If you don't require enabled auth, this process is more simple)

**Comment by hhorak** [comment]
@omron93 wrote:
> @mfojtik wrote:

>> What is the difference, it seems like we contain the same set of variables.

> RHSCL MongoDB by default (without any variables) starts without auth. You can enable it with MONGODB_AUTH=true (no user creation - useful with mounted database with initialized users) or by setting MONGODB_ADMIN_PASSWORD=xxx (creates admin user and enable auth). Additionally to MONGODB_ADMIN_PASSWORD you can also specify one user you want to create...

> OpenShift mongodb24 aren't able to run without auth or enable auth without creating users (MONGODB_USER + MONGODB_PASSWORD). If you want to create new user you don't have to specify database in which to create it (username by default). Also you don't have to specify (MONGODB_ADMIN_PASSWORD) and mongodb generate admin password and print it to stdout.

If I'm not mistaken the RHSCL image should work for all cases the openshift needs. Adding some more use cases isn't something openshift should block us, even if they cannot use them.

The case with "generate admin password and print it to stdout" is a security issue and openshift images should actually fix it in case those images should be released eventually.

**Comment by omron93** [comment]
> The case with "generate admin password and print it to stdout" is a security issue and openshift images should actually fix it in case those images should be released eventually.

I am sorry, it is not printed in actual version. It was removed... So mongodb24 generates admin password (and creates admin user), but you don't know it. And user you have created don't have permissions to administrate database (so you can only use the collection you have specified during start).

The difference is, that openshift won't be able to start container only with MONGODB_USER + MONGODB_PASSWORD...

**Comment by mfojtik** [comment]
> Or it is also possible when you have mounted database (to persist container restart). And you can start container with some options, connect to it with mongo shell, run something, restart mogodb container with different options, connect with mongo shell and initiate replicaset. (If you don't require enabled auth, this process is more simple)

This all is not valid for cluster, where the container does not have state and can be shot and restarted on another node. We will support persistent VOLUME for database storage but not for configuration files.

Replication: Yeah, I went through that process here:  https://github.com/openshift/mongodb/blob/master/2.4/contrib/initiate_replica.sh (just to give you an example how customers will use the image to do replication on Atomic/OpenShift).

**Comment by mfojtik** [comment]
@hhorak where do we print admin password to stdout?

**Comment by mfojtik** [comment]
@omron93 In OpenShift you can **set** the admin password when you creating the MongoDB template. It is an env variable and if you provide the value, then you own and know the password ;-)

Also the password is available for other containers as environment variable, so they can consume it without knowing the actual value.

Additionally, most of the use-cases we have does not require admin to login and users will just stick to the 'default' user. The 'admin' account will be used only to perform administration tasks, such as replication setup, add more users, etc...

**Comment by omron93** [comment]
>k where do we print admin password to stdout?

I am sorry, it is not valid now. You have removed this (echo "=> Creating an admin user with a ${MONGODB_ADMIN_PASSWORD} password in MongoDB") when you merged replication support...

**Comment by mfojtik** [comment]
@omron93 is there problem with generating the admin password when it is not set? Typical users don't need that password or admin account.

**Comment by mfojtik** [comment]
@omron93 I see the problem with --auth. We require auth because when you start the container in cluster you can't just restart it with to pickup '--auth'.

The approach we use now is that we start mongo (binding on localhost/socket) **without** auth. Then create users, then stop mongo and start **with** auth. During the users creation, we touch the ".users-created" file in data directory, so if you start another mongo container it knows that the users are already created.

**Comment by hhorak** [comment]
@mfojtik wrote:
> @omron93 I see the problem with --auth. We require auth because when you start the container in cluster you can't just restart it with to pickup '--auth'.

So then set `MONGODB_AUTH=true` in openshift and you're fine.

**Comment by mfojtik** [comment]
@hhorak don't understand why it can't be default? you most likely don't want to run Mongo without auth in production, correct?

**Comment by mfojtik** [comment]
@hhorak you have to disable auth just in one specific case, when you bootstrap a new mongo instance. Does that qualify to make this default behavior of the image?

**Comment by omron93** [comment]
> This all is not valid for cluster, where the container does not have state and can be shot and restarted on another node. We will support persistent VOLUME for database storage but not for configuration files.

I know. But this is not about config files. For example replication: start one container, normally create users and stop. With the same database (with created users) run container with specified --replSet and --keyFile and also 2 container with same --replSet + keyFile. Connect the first one and initiate replica set...

> Replication: Yeah, I went through that process here: https://github.com/openshift/mongodb/blob/master/2.4/contrib/initiate_replica.sh (just to give you an example how customers will use the image to do replication on Atomic/OpenShift). 

And these scripts can be mounted into container or included in a new layer.

>In OpenShift you can set the admin password when you creating the MongoDB template. It is an env variable and if you provide the value, then you own and know the password ;-)

>Also the password is available for other containers as environment variable, so they can consume it without knowing the actual value.

>Additionally, most of the use-cases we have does not require admin to login and users will just stick to the 'default' user. The 'admin' account will be used only to perform administration tasks, such as replication setup, add more users, etc... 

But there is a problem if you have mounted volume for database. When you stop this container you don't know the generated password! And this admin user of this database still have it. So then you have to start container without auth, change it to something and again start with enabled auth. For me it is better to have to specify it.

>The approach we use now is that we start mongo (binding on localhost/socket) without auth. Then create users, then stop mongo and start with auth. During the users creation, we touch the ".users-created" file in data directory, so if you start another mongo container it knows that the users are already created.

But it don't know what user and in which database. User should know best if users are created, what, where,... and then user can specify if he wants to enable auth, create another user, do nothing,...

> don't understand why it can't be default? you most likely don't want to run Mongo without auth in production, correct?

Maybe. So why upstream don't enable it by default? Even rpms offered to RHEL users (by upstream) have it disabled by default. (http://docs.mongodb.org/manual/tutorial/install-mongodb-on-red-hat/)

OpenShift is not able to have MONGODB_AUTH=true predefined in template?

**Comment by mfojtik** [comment]
> But there is a problem if you have mounted volume for database. When you stop this container you don't know the generated password! And this admin user of this database still have it. So then you have to start container without auth, change it to something and again start with enabled auth. For me it is better to have to specify it.

OK, I think your point is valid. We can generate the admin password inside the OpenShift template, so it will always be set. I'm fine with requiring the admin password for now. @jhadvig any concerns here?

> But it don't know what user and in which database. User should know best if users are created, what, where,... and then user can specify if he wants to enable auth, create another user, do nothing,...

You know that as they are environment variables that we require to be set (MONGODB_USER and MONGODB_PASSWORD and MONGODB_DATABASE). 

> Maybe. So why upstream don't enable it by default? Even rpms offered to RHEL users (by upstream) have it disabled by default. (http://docs.mongodb.org/manual/tutorial/install-mongodb-on-red-hat/)

What is upstream here? Docker image != RPM. We are going to provide bootstraping to our users so they don't have to perform these administrative tasks. The default should be MONGODB_AUTH=true and you can set that to MONGODB_AUTH=false if you want to perform the administrative tasks you described. 

If we are going to always use MONGODB_AUTH=true in OpenShift that it will cause inconsistent user experience if they want to use the image outside OpenShift, right? We don't want that.

**Comment by hhorak** [comment]
@mfojtik wrote:
> we require to be set (MONGODB_USER and MONGODB_PASSWORD and MONGODB_DATABASE)

please do so in openshift, but it doesn't mean the images need to require them as well.

**Comment by mfojtik** [comment]
@hhorak if we do that in openshift then users will get different experience in OpenShift and outside OpenShift. is that what we want?

**Comment by hhorak** [comment]
@mfojtik wrote:
> @hhorak if we do that in openshift then users will get different experience in OpenShift and outside OpenShift. is that what we want?

Yes, in this case it is what I'd like to see. "something is not possible in openshift" doesn't mean it cannot be possible in non-openshift use case.

**Comment by mfojtik** [comment]
@hhorak if so, why not make it default to work with openshift and provide option to work outside openshift? (seems this falls into the PM thread ;-)

**Comment by praiskup** [comment]
@mfojtik is this relevant to PostgreSQL too?  Because I think it is the same issue.
We allow you to specify that variables, but it is not requirement.  You require that to
be set in your UI, or?

**Comment by omron93** [comment]
>>  But it don't know what user and in which database. User should know best if users are created, what, where,... and then user can specify if he wants to enable auth, create another user, do nothing,...

>You know that as they are environment variables that we require to be set (MONGODB_USER and MONGODB_PASSWORD and MONGODB_DATABASE). 

I wanted to be able use container similarly:
You can mount already initialized database into container. And (for MongoDB) in this database there is collection with MongoDB users. You stop container. And after some time you want to run container with this database again.
You can one of there options:
- start container without auth (everything OK)
- enable auth (MONGODB_AUTH) and use users already created
- enable auth and also be able to create user, which is not in database yet

But I realized that the third point is not possible now, because it is not possible to create admin user twice.
So it should be able to enable auth only by specifying USER+PASS+DATABASE.

One question. Is acceptable to us to enable users to create only one database user without right to manage database?

**Comment by hhorak** [comment]
@mfojtik wrote:
> @hhorak if so, why not make it default to work with openshift and provide option to work outside openshift? (seems this falls into the PM thread ;-)

Just for the understanding as a user -- the UI requires it, not the image, which seems totally fine to me. Generally we should make sure users **may** use the images outside of openshift in the same way as in the openshift, but we cannot **require** it in all cases -- that's what we should make clear with PM.

**Comment by omron93** [comment]
>you have to disable auth just in one specific case, when you bootstrap a new mongo instance. Does that qualify to make this default behavior of the image?

To note, if we allow user to enable auth without creating admin, running container without auth for example 
will be used to create admin user into already initialized database.

**Comment by hhorak** [comment]
@mfojtik wrote:
> @hhorak if so, why not make it default to work with openshift and provide option to work outside openshift? (seems this falls into the PM thread ;-)

It'd be funny to ship images as ready as possible but then require to set environment variable to **just work** everytime.

**Comment by mfojtik** [comment]
@hhorak without env vars we basically saying that users should go bootstrap the container themself while in openshift we want users to have minimal effort with bootstraping them.

**Comment by mfojtik** [comment]
@hhorak if we do you approach as default (DIY) then when users come to OpenShift they will find different user experience. Where if we do OpenShift approach as default then users will find that outside OpenShift as well (and they will also still be able to run without env vars)

**Comment by praiskup** [comment]
@mfojtik, it sounds like topic for yet another issue for best practices.

Can you please start to be constructive?  Why we should fail for users
which simply do not want to have `DBPREFIX_{USER,DBNAME,PASSWORD}` specified?

**Comment by omron93** [comment]
@mfojtik And why openshift want to hide settings to users? Why doesn't it simply predefine this into "some template" and doesn't say this during creating database ("# These three variables are default by openshift. Do not change unless you know what you are doing"). Then users will also have the same user experience...

**Comment by hhorak** [comment]
@mfojtik wrote:
> @hhorak without env vars we basically saying that users should go bootstrap the container themself while in openshift we want users to have minimal effort with bootstraping them.

The effort is the same -- provide the initial authentication if they want. If not, fine. They can still use it very nicely.

**Comment by mfojtik** [comment]
@praiskup @hhorak because without those env vars set, how you want to use the image inside Kubernetes? Just start the image and then what? How you will connect to the database from other Pods?  How you will know what user/pass/db settings you should use? Or you expect users to do all this steps manually?

**Comment by mfojtik** [comment]
@praiskup then outside OpenShift, you start the image without anything, assuming that users will be smart enough to boostrap the database themself (which might be the case), however when they will move to OpenShift they will get completely different user-experience, where we do all these tasks for them. 

@praiskup read my last email in PM thread. Problem is that what you are proposing is perfectly fine for IaaS/hosted/DIY environment and I like it, but it is not valid for PaaS case where you **don't** want users to run all this steps manually.

**Comment by praiskup** [comment]
@mfojtik, your environment is probably weak enough to allow users to start with
multiple users, databases, ACLs, and so on.  Is that reason why we should block
you to implement such features in Kubernetes in future?

**Comment by mfojtik** [comment]
@praiskup additionally, trying to fit these two fundamentally different concepts into one image cause all this never-ending threads. I can say the same about your constructionist, right?  Instead of helping users to make the images usable out-of-box, you just saying that you want to delegate all bootstraping and configuration on them as they know what they want the best. Which is caused by not understanding the Docker principles and why we are doing Docker images....

**Comment by mfojtik** [comment]
@praiskup our 'weak' environment is what our users want and customers are paying for. There are no users/customers requests to implement these features as far as I know.

**Comment by hhorak** [comment]
@mfojtik not everybody will use the image inside kubernetes.
if users are not smart enough to bootstrap db, then they will use the env. variables, that's what doc is for.

"make the images usable out-of-box" is exactly the way I'd like to see it, without need to specify a database..

**Comment by hhorak** [comment]
I think we're really stuck here, let's wait for the pm..

**Comment by mfojtik** [comment]
@hhorak @praiskup 

Having these env vars not set when you start the image with attached  storage that contain existing database means that when you link this container with another container you won't be able to connect to the database from the linked container as you will not have the authentication information available there.

You can still bypass the check if they are set when you override the default command, right? "docker run mysql mysqld".

**Comment by praiskup** [comment]
@mfojtik Michale, I did not get that question.  Like you have pre-initilaized
the database already?  E.g. from previous container run?

Then yes, there is no need to specify `POSTGRESQL_ADMIN_PASSWORD`
from the container perspective so it is not required by container itself.  It is fair
point you will require that however - and use that for linking.  Anyway, you know
best what the content of POSTGRESQL_ADMIN_PASSWORD should be, container
has no idea to say it is correct/wrong content for the second run..

**Comment by praiskup** [comment]
> You can still bypass the check if they are set when you override the default
> command, right? "docker run mysql mysqld".

Sorry, I did not get what you mean by this?  Which check actually?

**Comment by mfojtik** [comment]
@praiskup i'm not talking about the POSTGRESQL_ADMIN_PASSWORD but about the USER/PASSWORD.
I think the problem we are trying to solve here was that where in OpenShift want to have them mandatory, you don't want to require them.
My use-case was that when you don't provide them (make them mandatory) then it breaks the linking between containers as the container to where you are linking the mysql/database container to won't be able to pickup POSTGRESQL_USER/PASSWORD and connect to database. Does it make sense?

**Comment by mfojtik** [comment]
@hhorak @praiskup we talked with Pavel on IRC and we agreed that following approach would be feasible for both sides:

If you set `POSTGRESQL_SOCKET` then _USER, _PASSWORD and _ADMIN_PASSWORD are all optional as you want to do local linking via socket which can't be done via kubernetes today (afaik).

If you don't set this option, then either:

1) `_USER` and `_PASSWORD` are both mandatory
2) `_ADMIN_PASSWORD` is mandatory

By setting 1) and 2) together, you get _USER and _PASSWORD create as well as _ADMIN_PASSWORD. If you don't set any option them, the image will fail to run, displaying the `container-usage`.

@hhorak would that be doable for other databases? This works for OpenShift (@jhadvig, @dmcphers) as it does not break what we have currently (except that we don't support **only** having _ADMIN_PASSWORD set, but I think it is a reasonable compromise and covers the case when you don't want to create the default user).

**Comment by praiskup** [comment]
@hhorak, @mfojtik, sorry for the delay, I had to travel a bit.

I don't have a strong opinion here.  Making something explicit is not bad
IMO.  So from my POV its pretty fine.  Yes, running 'container-usage' sounds
like good compromise ;).  But I need to hear @hhorak voice.

We also talked about the generic approach which not something would hurt
OpenShift users, just not allowed in UI ('POSTGRESQL_CONFIG'), and it
is not going to be used.

Also, as the environment variables will be documented in product
documentation, and tested by OpenShift QA, we are free to implement those
variables in a '10-openshift.sh' hook - always available, however.  @mfojtik, is
that correct?

Honzo, are you OK with those ^^?

**Comment by praiskup** [comment]
Also documentation of variables like POSTGRESQL_MAX_CONNECTIONS in code,
README.md, container-usage is not that useful, because that variables are
something like helper for OpenShift UI, where all the variables will be properly
described/documented.  That seems to be reasonable.  @mfojtik, correct me if I'm
wrong.

**Comment by mfojtik** [comment]
@praiskup as they will part of the image, I think we can mention them in your README.md (as we don't want to hide them). It can be documented as "OpenShift" or "PaaS" variables. We can keep container-usage minimal, alternatively point to the OpenShift product documentation:

http://docs.openshift.org/latest/using_images/db_images/mysql.html#environment-variables

**Comment by mfojtik** [comment]
@praiskup and yes, what you wrote about openshift hook is correct.

**Comment by mfojtik** [comment]
@danmcp @bparees @jhadvig is that reasonable? (https://github.com/sclorg/rhscl2dockerfile/issues/21#issuecomment-104686817)

**Comment by praiskup** [comment]
@mfojtik wrote
> @praiskup as they will be part of the image, I think we can mention them in
> your README.md (as we don't want to hide them).

"We" can mention them in our image :).  Ok, Honza should comment probably
but I am not going to block that if you agree that "mentioning" is not
documentation (and thus duplicate of upstream work).

Yet another on-topic idea, do you really insist on naming like like
"POSTGRESQL_CONFIG_NAME", or would that be ok to find some OpenShift (or cloud
lets say) prefix?  To not waste the 'POSTGRESQL_' namespace?

**Comment by praiskup** [comment]
Having it namespaced in OSE_ or something like that would allow us to
easily move to better configuration options in future (already discussed)
without breaking API.

Also, I don't think that POSTGRESQL_USER is equivalent semantically to
POSTGRESQL_SHARED_BUFFERS.

**Comment by danmcp** [comment]
@mfojtik @praiskup I don't think it is acceptable to have one image with two totally different experiences recommended.  Both cases should have the same goals.  They all want tunability, configurability, scale/clustering, logging, metrics, debugability, etc.  And we don't want to have people set this up with OpenShift specific variables.  We just need to get to a common ground where we can deliver that one experience in all cases.

**Comment by omron93** [comment]
>If you set POSTGRESQL_SOCKET then _USER, _PASSWORD and _ADMIN_PASSWORD are all optional as you want to do local linking via socket which can't be done via kubernetes today (afaik).

And what about MongoDB? There also should be an option to create unix socket?

>By setting 1) and 2) together, you get _USER and _PASSWORD create as well as _ADMIN_PASSWORD. If you don't set any option then, the image will fail to run, displaying the container-usage.

And what about _DATABASE? Should it be mandatory? Or by default same as the _USER?...

And what to do, if there is already created _USER in mounted database? I hope it is suitable to simply print "Failed to create user" and print container-usage... ?

Any option to disable auth? (MONGODB_AUTH=false)?

**Comment by hhorak** [comment]
We've just chatted off-github, but the plan as for today is to copy functionality from https://github.com/openshift/mongodb/tree/master/2.4 as it is and not add anything more, at least as for now.

**Comment by omron93** [comment]
One question to be sure - "copy functionality" includes tests and also replication?

**Comment by mfojtik** [comment]
@omron93 we don't require to have replication there (it is part of the entrypoint currently, but it is optional...) @bparees do we want to get rid of that for release or keep the replication there?

**Comment by omron93** [comment]
@mfojtik  And what about run-mongos.sh script? (it starts mongos - sharding server)

OpenShift does not include it. It is not working out-of-the box (you are not able to only run container with CMD run-mongos - there should be specified at least one configuration server), but it should be necessary (or there could be a mounted script into run-mongod which completely change exec binary) to enable sharding. Another possible feature - similar to replication.

**Comment by omron93** [comment]
> Also the password is available for other containers as environment variable, so they can consume it without knowing the actual value.

@mfojtik @jhadvig During changing MongoDB docker image to fit OpenShift. I've founded out that this is not true about generated admin password. It is not available as environment variable. See:
"local admin_pwd=${MONGODB_ADMIN_PASSWORD:-$(< /dev/urandom tr -dc _A-Z-a-z-0-9 | head -c16)}"
Should this be same in RHSCL MongoDB image?

**Comment by mfojtik** [comment]
@omron93 I think the reasoning for this was that we don't want to make the ENV vars exported for the running process in case of security bug... but I will make @jhadvig or @mnagy confirm that.

**Comment by mnagy** [comment]
@omron93 @mfojtik Yes, this is intentionally unset in the entrypoint script before the final execution of mongo.

**Comment by omron93** [comment]
@mfojtik OK. I've mentioned this, because you wrote arguments for generating admin password:
>In OpenShift you can set the admin password when you creating the MongoDB template. It is an env variable and if you provide the value, then you own and know the password ;-)

>Also the password is available for other containers as environment variable, so they can consume it without knowing the actual value.

>Additionally, most of the use-cases we have does not require admin to login and users will just stick to the 'default' user. The 'admin' account will be used only to perform administration tasks, such as replication setup, add more users, etc... 

But because of having "local admin_pwd" this password won't be available for administration task in container. Isn't this correct?

Which environmental variables should be available for other containers?

**Comment by mfojtik** [comment]
@omron93 if you set this environment variable in kube template for given container it will be still available when you do `docker exec` (or `osc exec`) it just wont be available for the process that runs as PID 1 (or at least that is my understanding).

**Comment by omron93** [comment]
@mfojtik But if you don't set it MongoDB container generate this admin password. And you are not able to get this password. (even `docker exec` not) Am I wrong? So you can't use admin user for administration tasks...

And I understand `unset MONGODB_USER MONGODB_PASSWORD MONGODB_DATABASE MONGODB_ADMIN_PASSWORD` before  `exec mongod`, but this is not what I am asking about...

**Comment by omron93** [comment]
And if you set MONGODB_ADMIN_PASSWORD and it is unset before `exec mongod`. You still can get this admin password with `docker exec`...

I am asking about `local` for generated admin password. Because if you generate admin password (MONGODB_ADMIN_PASSWORD is not set), admin user is totally unusable...

**Comment by mfojtik** [comment]
@omron93 you're correct, this works for other databases (as you can connect without authentication using local socket) but not in MongoDB. @mnagy can you please make the MONGO_ADMIN_PASSWORD required? @bparees CC

@omron93 thanks for reporting this.

**Comment by omron93** [comment]
@mfojtik Unix socket support provided by MongoDB is not usable in the same way as for other databases?

**Comment by mfojtik** [comment]
@omron93 it is possible to use UNIX socket in mongo if mongo is started with `--auth` without providing any authentication?

**Comment by omron93** [comment]
@mfojtik I've tried it and it is not possible. So MONGODB_ADMIN_PASSWORD should be required?

**Comment by mfojtik** [comment]
@omron93 i believe so... we tried to make this work the same as for other database images, but I think it is not possible for mongo.
